npep/npep-122.md

```diff
@@ -0,0 +1,194 @@
+
+The wording used here has lead to much confusion. Specifically, "By default" part actually means this policy should 
+be at the BANP priority, and solves the "Engineering org vs Marketing org" use case where you want to keep orgs 
+from accidentally interfering with each other, but still want to allow them to communicate if needed. 
+
+In reality, the CR attached to this story defines strict tenancy, which should be a separate use case (see next section).
+
+To make the use case more obvious we can add some details to it like the following
+
+    As a cluster admin, I want to build tenants in my cluster that are isolated from each other by default.
+    This should prevent tenants from interfering with each other by accident, but may be overridden by namespace
+    owners when needed.
+    Tenancy may be modeled as 1:1, where 1 tenant is mapped to a single Namespace, or 1:n, where a single tenant 
+    may own more than 1 Namespace.
+
+#### Story 4.2: Create and Isolate multiple tenants in a cluster, strict isolation
+
+Strict tenancy is the "Coke vs Pepsi" sort of thing where you want each tenant to feel like it has its own cluster, 
+and be totally independent of the other tenants. We can write it down like this
+
+    As a cluster admin, I want to build tenants in my cluster that are isolated from each other and this can't
+    be overridden. This policy should make every tenant completely independent and isolated from another tenants.
+    Tenancy may be modeled as 1:1, where 1 tenant is mapped to a single Namespace, or 1:n, where a single tenant 
+    may own more than 1 Namespace.
+
+#### Story 4.3: Allow internal connections for tenants
+    
+    As a cluster admin, I want to build tenants in my cluster and always allow connections inside one tenant.
+    At the same time I want to setup an overridable deny-all policy to protect namespaces by default.
+    This policy should make sure internal connectivity for a tenant is always allowed, in case there are
+    lower-priority deny rules.
+
+#### Story 4.4: Tenants interaction with (B)ANP
+
+    As a cluster admin, I want to be able to setup policies with higher and lower priority than tenancy policy.
+    Higher priority to allow/deny some inter-tenant connections, for example, allow ingress traffic from the
+    monitoring namespace without including that namespace to every tenant. Lower priority to allow/deny
+    connections that are not a part of tenant policy, for example, allow internal tenant connections with tenancy
+    and deny-all policy with lower priority.
+
+#### Story 4.5: Ensure every namespace in the cluster is a part of some tenant
+
+    As a cluster admin, I want to implement security zones tenants, and deny connections between different zones.
+    I want to use a label e.g. `security-zone` to build tenants, and I want to make sure every namespace in the cluster
+    is a part of some tenant/security zone. This way, I can be sure that even if the `security-zone` 
```

Comment:
Yes it is similar to 4.2, but this one stresses the "none" tenant as a requirement. Currently, we don't have a way to create "none" tenant, that is why I added an extra story to explain why this functionality may be needed explicitly

Reply:
Hrm this seems dangerous to me... almost like the Implicit deny provided by network policy...

i.e we're saying here that when there's no tenants defined nothing happens... but the moment there's at least on tenant ALL non selected workloads are/ can be implicitly selected like a tenant now? 

i.e I think we need to stay away from use cases where rules are applied to workloads even if they aren't explicitly selected by a policy  

maybe I'm overthinking it though 🤷

Reply:
I completely agree with your statements above. I think the confusion though comes from whether this is a "use case" v.s. "api semantics clarification". IMO, we should actually provide sample tenant-rule policy which only selects namespaces that contains the tenancy label(s) as a best practice. In addition, we clarify that if "extra" namespaces are selected, what's the expected behavior, and _maybe_ that is not recommended. But the other way around does not make sense to me: because I want to have every namespace in the cluster strictly associated with a tenant, let's select all namespaces in the policy while I could have just selected the "tenant namespaces" that I care about.